### PR TITLE
StyleGAN2 ADNI brain dataset - 47903151

### PR DESCRIPTION
**Pull request by firefried-fries**
### Description:
This is an implementation of StyleGAN2 image generation model on the ANDI brain dataset, with umap plots comparing style embedding of train models on AD and NI classes separately. 

#### Main Files
modules.py - main files that defines the model and a few helper functions.
train.py - main file for training model. It is used to train and save the model.
predict.py - a demonstration of the model usage. Loads a given model and plots fake images from generator against real images. If --plot_umap is specified to be True and --AD_dir, --NC_dir are not empty and are directories of saved model, it will attempt to plot the style embeddings of the two models.
dataset.py - have a CustomImageFolder class that allows grabbing only specified class, and get_loader function which returns the data loader.
constants.py - stores the constants and hyper parameters. 
README.md - A description of the project with sample usages and training plots.
#### Folders
model - contains a trained model on both AD and NC images
AD - contains a trained model on only the AD class
NC - contains a trained model on only the NC class
training - contains the training loss plot
umap plot - contain the umap plot
fake_vs_real - generated fake images at different epoches against real images

Thanks to all the course staff! - student 47903151

PS. This is my first time doing a project and honestly it's fun, especially the fight over rangpur resources, I would constantly squeue and stare at how much more people are ahead of me at 3am (average 4 hours wait lol :sob:)

### CODE REVIEW

**Comment by Uto0op** [comment]
Recognition problem: StyleGAN on ADNI ---- Hard Difficulty

1. Algorithm solves the problem: (5 Marks)

2. Algorithm demonstrated and functions: (3 Marks)

4. Design: (1 Mark)
Highly modularized design

5. Commenting:
Sufficient docstrings but comments could be more specific -0.5 (1 Mark)

6. Algorithm above Normal Difficulty: (5 Marks)

7. Algorithm is Hard difficulty: (5 Marks)

**Suggestions**:
- It's good to record training results, losses and UMAP progressively, you might also want to add some brief justifications on the losses;
- It seems the UMAP was embedded on the style space of the AD and NC models separately and plotted together, while the intention of this task is to have the embedding on the "general model"'s style space, which is what they did in the reference paper. This is fine, however, since the result agrees with the paper, through a different approach;
- Good project.

**Comment by yexincheng** [comment]
**This is an initial inspection, no action is required at this point**
1. **Solves problem**: Could have more explanation of the network. The generated images are realistic. The training plots were provided but the plot titles are less informative; I am confused about their difference, but they all look reasonable. Nice UMAP.
2. **Implementation**: Data augmentation implemented. It is hyperparameters in the constants.py which can be tuned; constants is not a proper name. Nice usage of ArgParse. Could have a function for training loop before program entry.
3. **Design**: Good
4. **Comment**: More commented are required for modules.py; almost no comments for now. Others are good.
5. **Difficulty**: Hard

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting missing. Sufficient docstrings but comments could be more specific.</td><td>-1</td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. StyleGAN (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Some/Adequate Meaningful commit messages. Abstract or minimal.</td><td>-1</td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). ON time.</td><td></td></tr>
	<tr><td></td><td>Feedback action required: Feedback marks possible +2 if the requested changes are made. Remove model files and clean up for merge.</td><td>-2</td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-4</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by gayanku** [comment]
Feedback marks possible +2 if the requested changes are made (see above).

**Comment by hanemma7moud** [comment]
Yixin Hu

**Comment by firefried-fries** [comment]
Added more docstring to module.py
Accidentally forgot to change commit message, no mistake with umap plot was corrected

**Comment by shakes76** [comment]
Cannot merge because of model files, please remove for merging. Doesnt affect grade. +2 for feedback
